backends/k8s/client.go

```diff
@@ -106,212 +117,176 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	} else {
+		kvps = append(kvps, peers...)
 	}
-	return waitIndex, nil
+
+	// Populate the vars map from the KVPairs.
+	c.populateFromKVPairs(kvps, vars)
+
+	return nil
 }
 
-// buildTPRClient builds a RESTClient configured to interact with Calico ThirdPartyResources.
-func buildTPRClient(baseConfig *rest.Config) (*rest.RESTClient, error) {
-	// Generate config using the base config.
-	cfg := baseConfig
-	cfg.GroupVersion = &schema.GroupVersion{
-		Group:   "projectcalico.org",
-		Version: "v1",
+// populateFromKVPairs populates the vars KV map from the supplied set of
+// KVPairs.  This uses the libcalico-go compat module and serialization functions
+// to write out the KVPairs in etcdv2 format.  This works in conjunction with the
+// etcdVarClient defined below which provides a "mock" etcd backend which actually
+// just writes out data to the vars map.
+func (c *Client) populateFromKVPairs(kvps []*model.KVPair, vars map[string]string) {
+	// Create a etcdVarClient to write the KVP results in the vars map, using the
+	// compat adaptor to write the values in etcdv2 format.
+	client := compat.NewAdaptor(&etcdVarClient{vars: vars})
+	for _, kvp := range kvps {
+		client.Apply(kvp)
 	}
-	cfg.APIPath = "/apis"
-	cfg.ContentType = runtime.ContentTypeJSON
-	cfg.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: clientapi.Codecs}
+}
+
+// etcdVarClient implements the libcalico-go backend api.Client interface.  It is used to
+// write the KVPairs retrieved from the Kubernetes datastore driver into the KV map
+// using etcdv2 naming scheme.
+type etcdVarClient struct {
+	vars map[string]string
+}
+
+func (c *etcdVarClient) Create(kvp *model.KVPair) (*model.KVPair, error) {
+	log.Fatal("Create should not be invoked")
+	return nil, nil
+}
 
-	cli, err := rest.RESTClientFor(cfg)
+func (c *etcdVarClient) Update(kvp *model.KVPair) (*model.KVPair, error) {
+	log.Fatal("Update should not be invoked")
+	return nil, nil
+}
+
+func (c *etcdVarClient) Apply(kvp *model.KVPair) (*model.KVPair, error) {
+	path, err := model.KeyToDefaultPath(kvp.Key)
 	if err != nil {
+		log.Error("Unable to create path from Key: %s", kvp.Key)
 		return nil, err
 	}
+	value, err := model.SerializeValue(kvp)
+	if err != nil {
+		log.Error("Unable to serialize value: %s", kvp.Key)
+		return nil, err
+	}
+	c.vars[path] = string(value)
+	return kvp, nil
+}
 
-	// We also need to register resources.
-	schemeBuilder := runtime.NewSchemeBuilder(
-		func(scheme *runtime.Scheme) error {
-			scheme.AddKnownTypes(
-				*cfg.GroupVersion,
-				&thirdparty.GlobalConfig{},
-				&thirdparty.GlobalConfigList{},
-				&thirdparty.IpPool{},
-				&thirdparty.IpPoolList{},
-				&kapiv1.ListOptions{},
-				&kapiv1.DeleteOptions{},
-			)
-			return nil
-		})
-	schemeBuilder.AddToScheme(clientapi.Scheme)
-
-	return cli, nil
+func (c *etcdVarClient) Delete(kvp *model.KVPair) error {
+	// Delete may be invoked as part of the multi-key resources, but since we start
```

Comment:
You mean that Delete is called as part of the compat adapter's Apply processing even though we're not calling it directly?